Fix TOCTOU vulnerability in unused_port module

sashaodessa opened the pull request:
Replaces unsafe port-finding functions with secure bound socket APIs to eliminate race conditions.

### Problem
The existing `unused_tcp*_port()` and `unused_udp*_port()` functions had a classic **Time-of-Check-Time-of-Use (TOCTOU)** vulnerability:
- Functions would bind to port 0, get the assigned port number, then immediately close the socket
- Between getting the port number and later binding to it, another process could claim the same port
- The LRU cache provided false security and didn't prevent the race condition

### Solution
Added new safe APIs that return already-bound sockets:
- `bind_tcp4_any()` / `bind_tcp6_any()` → returns `TcpListener`
- `bind_udp4_any()` / `bind_udp6_any()` → returns `UdpSocket`

jimmygchen committed:
Add co-author to mergify commits (#7993)

* Add co-author to mergify commits.

* Remove unnecessary pull request rules from mergify config.

* Revert automation removals

sashaodessa committed:
fix TOCTOU vulnerability in unused_port module

michaelsproul commented:
Thanks!

We are aware of this issue and had been phasing these functions out in favour of using 0 ports explicitly. The hard part is replacing the uses of the old functions in code that assumes it can determine an unused port without binding a socket, e.g. here:

https://github.com/sigp/lighthouse/blob/fd10b632740e441354c4547a12727c29dd288239/lighthouse/tests/beacon_node.rs#L955-L970

If you would be interested in coming up with clever fixes for these cases, that would be extremely helpful. That would then allow us to delete rather than deprecate the unsafe functions.

michaelsproul commented:
Most of the uses are in tests, but it's still important for us to get rid of them (if possible), as they sometimes contribute to CI flakiness.

daveth3t3chg33k commented on beacon_node/src/config.rs:
```diff
@@ -1058,24 +1055,21 @@ pub fn parse_listening_addresses(cli_args: &ArgMatches) -> Result<ListenAddress,
-                .or(maybe_quic_port)
-                .unwrap_or(if tcp_port == 0 { 0 } else { tcp_port + 1 });
+            let quic_port = if use_zero_ports {
+                0
```
Extrapolate the function to classes

daveth3t3chg33k reviewed (commented):
Extrapolate the tcp function into the udp main class

sashaodessa commented:
@daveth3t3chg33k

michaelsproul commented:
@sashaodessa Can you also please rebase on `unstable`? All PRs need to be merged to `unstable`

sashaodessa commented:
@michaelsproul

michaelsproul commented:
Looks good at a glance, I've requested a review from some our networking experts on the changes that impact prod

AgeManning commented on .github/mergify.yml:
```diff
@@ -105,6 +105,10 @@ queue_rules:
 
 
         {{ body | get_section("## Proposed Changes", "") }}
+      
+      {% for commit in commits | unique(attribute='email_author') %}
+      Co-Authored-By: {{ commit.author }} <{{ commit.email_author }}>
+      {% endfor %}
```
Are these changes relevant?

michaelsproul replied:
No this is some leftover stuff from `stable` that hasn't been backmerged to `unstable` yet (see https://github.com/sigp/lighthouse/commits/stable/), it should be deleted from this PR.

jimmygchen committed:
Update engine methods in notifier (#8038)

Fulu uses `getPayloadV5`, this PR updates the notifier logging prior to the fork.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

AgeManning reviewed (commented):
These changes look good to me. 

Agree a future PR to remove deprecation would be nice.

michaelsproul commented:
I think we can probably just delete the deprecated functions in this PR, AFAICT all (?) usages have been removed

michaelsproul committed:
Another check to prevent duplicate block imports (#8050)

Attempt to address performance issues caused by importing the same block multiple times.


  - Check fork choice "after" obtaining the fork choice write lock in `BeaconChain::import_block`. We actually use an upgradable read lock, but this is semantically equivalent (the upgradable read has the advantage of not excluding regular reads).

The hope is that this change has several benefits:

1. By preventing duplicate block imports we save time repeating work inside `import_block` that is unnecessary, e.g. writing the state to disk. Although the store itself now takes some measures to avoid re-writing diffs, it is even better if we avoid a disk write entirely.
2. By returning `DuplicateFullyImported`, we reduce some duplicated work downstream. E.g. if multiple threads importing columns trigger `import_block`, now only _one_ of them will get a notification of the block import completing successfully, and only this one will run `recompute_head`. This should help avoid a situation where multiple beacon processor workers are consumed by threads blocking on the `recompute_head_lock`. However, a similar block-fest is still possible with the upgradable fork choice lock (a large number of threads can be blocked waiting for the first thread to complete block import).


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

eserilev committed:
Remove column reconstruction when processing rpc requests (#8051)

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

eserilev committed:
Reduce reconstruction queue capacity (#8053)

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

eserilev committed:
Add instrumentation to `recompute_head_at_slot` (#8049)

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

jimmygchen committed:
Enable reconstruction for nodes custodying more than 50% of columns and instrument tracing (#8052)

Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Jimmy Chen <jimmy@sigmaprime.io>

jking-aus committed:
chore: Bump discv5 and remove generic DefaultProtocolId in metrics (#8056)

Bump discv5 version


Co-Authored-By: Josh King <josh@sigmaprime.io>

dapplion committed:
Don't penalize peers for extending ignored chains (#8042)

Lookup sync has a cache of block roots "failed_chains". If a peer triggers a lookup for a block or descendant of a root in failed_chains the lookup is dropped and the peer penalized. However blocks are inserted into failed_chains for a single reason:
- If a chain is longer than 32 blocks the lookup is dropped to prevent OOM risks. However the peer is not at fault, since discovering an unknown chain longer than 32 blocks is not malicious. We just drop the lookup to sync the blocks from range forward sync.

This discrepancy is probably an oversight when changing old code. Before we used to add blocks that failed too many times to process to that cache. However, we don't do that anymore. Adding a block that fails too many times to process is an optimization to save resources in rare cases where peers keep sending us invalid blocks. In case that happens, today we keep trying to process the block, downscoring the peers and eventually disconnecting them. _IF_ we found that optimization to be necessary we should merge this PR (_Stricter match of BlockError in lookup sync_) first. IMO we are fine without the failed_chains cache and the ignored_chains cache will be obsolete with [tree sync](https://github.com/sigp/lighthouse/issues/7678) as the OOM risk of long lookup chains does not exist anymore.

Closes https://github.com/sigp/lighthouse/issues/7577


  Rename `failed_chains` for `ignored_chains` and don't penalize peers that trigger lookups for those blocks


Co-Authored-By: dapplion <35266934+dapplion@users.noreply.github.com>

gitToki committed:
fix(rate_limiter): add missing prune calls for light client protocols (#8058)

Co-Authored-By: Jimmy Chen <jimmy@sigmaprime.io>

Co-Authored-By: gitToki <tokipro@proton.me>

jimmygchen committed:
Update issue template (#7938)

* Update issue template

* Delete old issue template

eserilev committed:
Prevent silently dropping cell proof chunks (#8023)

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

michaelsproul committed:
Fix reprocess queue memory leak (#8065)

Fix a memory leak in the reprocess queue.


  If the vec of attestation IDs for a block is never evicted from the reprocess queue by a `BlockImported` event, then it stays in the map forever consuming memory. The fix is to remove the entry when its last attestation times out. We do similarly for light client updates.

In practice this will only occur if there is a race between adding an attestation to the queue and processing the `BlockImported` event, or if there are attestations for block roots that we never import (e.g. random block roots, block roots of invalid blocks).


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

michaelsproul committed:
Add experimental complete-blob-backfill flag (#7751)

A different (and complementary) approach for:

- https://github.com/sigp/lighthouse/issues/5391


  This PR adds a flag to set the DA boundary to the Deneb fork. The effect of this change is that Lighthouse will try to backfill _all_ blobs.

Most peers do not have this data, but I'm thinking that combined with `trusted-peers` this could be quite effective.


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

jimmygchen committed:
Add release helper script to list PRs and breaking changes (#7737)

Output for 7.1.0 release:

```
#   Commit SHA    PR Number    Has backwards-incompat Label    PR Title
---  ------------  -----------  ------------------------------  --------------------------------------------
1    d5a03c9d86bf  6872         False                           Add more range sync tests (#6872)
2    ec2fe3812edc  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0-beta.0' into unstable
3    3992d6ba74c9  6862         False                           Fix misc PeerDAS todos (#6862)
4    d60388134d07  6928         False                           Add PeerDAS metrics to track subnets without peers (#6928)
5    431dd7c39828  6917         False                           Remove un-used batch sync error condition (#6917)
6    0055af56b685  6932         False                           Unsubscribe blob topics at Fulu fork (#6932)
7    6ab6eae40c0e  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0-beta.0' into unstable
8    193061ff7376  6634         False                           Use RpcSend on RPC::self_limiter::ready_requests (#6634)
9    e5e43ecd8129  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0' into unstable
10   b4be5141823f  7012         False                           Add spamoor_blob in network_params.yaml (#7012)
11   01df433dfd02  7021         False                           update codeowners, to be more specific (#7021)
12   60964fc7b530  6829         False                           Expose blst internals (#6829)
13   3fab6a2c0ba7  6866         False                           Block availability data enum (#6866)
14   6e11bddd4bd0  6947         False                           feat: adds CLI flags to delay publishing for edge case testing on PeerDAS devnets (#6947)
15   454c7d05c40b  7017         False                           Remove LC server config from HTTP API (#7017)
16   54b4150a6220  7030         False                           Add test flag to override `SYNC_TOLERANCE_EPOCHS` for range sync testing (#7030)
17   cf4104abe5e2  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0' into unstable
18   8a772520a50a  7034         False                           Cache validator registration only after successful publish (#7034)
19   1235d4480225  7048         False                           Remove `watch` (#7048)
20   3bc5f1f2a58b  7081         False                           Validator Registration ssz support (#7081)
21   b4e79edf2a09  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0' into unstable
22   8d1abce26ed5  6915         False                           Bump SSZ version for larger bitfield `SmallVec` (#6915)
23   1916a2ac5ad3  7020         False                           chore: update to rust-eth-kzg to 0.5.4 (#7020)
24   1a08e6f0a090  7109         False                           Remove duplicate sync_tolerance_epochs config (#7109)
25   f23f984f8575  7057         False                           switch to upstream gossipsub (#7057)
26   d60c24ef1cc0  6339         True                            Integrate tracing (#6339)
27   a6bdc474db01  6991         False                           Log range sync download errors (#6991)
28   574b204bdb39  6680         False                           decouple `eth2` from `store` and `lighthouse_network` (#6680)
29   c095a0a58feb  7130         False                           update gossipsub to the latest upstream revision (#7130)
30   5cda1641ea2f  7137         False                           Log `file appender` initialization errors properly (#7137)
31   d96123b02882  7149         False                           Remove unnecessary `filter_layer` in logger builder (#7149)
32   a1b1d7ae589f  7150         False                           Remove `discv5` logs from logfile output (#7150)
33   ca237652f1da  6998         False                           Track request IDs in RangeBlockComponentsRequest (#6998)
34   d323699fde01  7183         False                           Add missing `osaka-time` lcli param (#7183)
35   cbf1c04a1486  -            -                               [NO PR MATCH]: resolve merge conflicts between untstable and release-v7.0.0
36   2f37bf4de5e3  -            -                               [NO PR MATCH]: Fix more merge conflicts between unstable and release-v7.0.0
37   3f6c11db0eb6  6995         False                           Some updates to Lighthouse book (#6995)
38   9dce729cb6a0  7182         False                           Ensure sqlite and rusqlite are optional in `consensus/types` (#7182)
39   6f31d4434308  7033         False                           Remove CGC from data_availability checker (#7033)
40   ca8eaea11677  7169         True                            Remove `crit` as an option from the CLI entirely (#7169)
41   bde0f1ef0b29  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0' into unstable
42   fb7ec0d151d4  7112         False                           Change `genesis-state-url-timeout` (#7112)
43   4839ed620fa9  7168         False                           Tracing cleanup (#7168)
44   578db67755cb  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/release-v7.0.0' into backmerge-apr-2
45   80626e58d224  7244         False                           Attempt to fix flaky network tests (#7244)
46   d6cd049a453b  7238         False                           RPC RequestId Cleanup (#7238)
47   0e6da0fcafe2  -            -                               [NO PR MATCH]: Merge branch 'release-v7.0.0' into v7-backmerge
48   57abffcd997f  7240         False                           Disable log color when running in non-interactive mode (#7240)
49   6a75f24ab13e  7188         False                           Fix the `getBlobs` metric and ensure it is recorded promptly to prevent miscounts (#7188)
50   7cc64cab8352  6990         False                           Add missing error log and remove redundant id field from lookup logs (#6990)
51   591fb7df141d  -            -                               [NO PR MATCH]: Merge branch 'release-v7.0.0' into backmerge-for-openssl
52   e77fb01a063c  7265         False                           Remove CLI conflict for secrets-dir and datadir (#7265)
53   b5d40e3db06d  7256         False                           Align logs (#7256)
54   70850fe58d56  6744         True                            Drop head tracker for summaries DAG (#6744)
55   47a85cd1186d  7269         False                           Bump version to v7.1.0-beta.0 (not a release) (#7269)
56   e924264e17b8  7258         False                           Fullnodes to publish data columns from EL `getBlobs` (#7258)
57   759b0612b37f  7117         False                           Offloading KZG Proof Computation from the beacon node (#7117)
58   d96b73152e0e  7192         False                           Fix for #6296: Deterministic RNG in peer DAS publish block tests (#7192)
59   39eb8145f89e  -            -                               [NO PR MATCH]: Merge branch 'release-v7.0.0' into unstable
60   70f8ab9a6fc2  7309         False                           Add riscv64 build support (#7309)
61   be68dd24d05f  7281         False                           Fix wrong custody column count for lookup blocks (#7281)
62   08882c64cae5  6996         False                           Fix execution engine integration tests with latest geth version (#6996)
63   476f3a593c20  7161         False                           Add `MAX_BLOBS_PER_BLOCK_FULU` config (#7161)
64   c32569ab83bb  7225         False                           Restore HTTP API logging and add more metrics (#7225)
65   410af7c5f5dc  7279         False                           feat: update mainnet bootnodes (#7279)
66   80fe133d2c4c  7280         False                           Update Lighthouse Book for Electra features (#7280)
67   9f4b0cdc2855  7343         False                           Fix Kurtosis doppelganger CI (#7343)
68   e61e92b926d5  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/stable' into unstable
69   5527125f5e13  7340         False                           Fix GitHub releases page looks bad in GitHub dark theme (#7340)
70   c13e069c9c63  7324         False                           Revise logging when `queue is full` (#7324)
71   1dd37048b9d1  7346         False                           Enable cross-compiling for riscv64 architecture (#7346)
72   402a81cdd78e  7350         False                           Fix Kurtosis testnet (#7350)
73   1324d3d3c4c2  5923         False                           Delayed RPC Send Using Tokens (#5923)
74   6fad18644bbe  6747         False                           feat: presign for validator account (#6747)
75   2e2b0d2176e0  7351         False                           Revise consolidation info in Lighthouse book (#7351)
76   63a10eaaea62  6956         True                            Changing `boot_enr.yaml` to expect `bootstap_nodes.yaml` for pectra devnet  (#6956)
77   34a6c3a93029  6897         True                            vc: increase default gas limit (#6897)
78   94ccd7608ea8  6653         False                           Add documentation for VC API `/lighthouse/beacon/health` (#6653)
79   9779b4ba2c04  7326         False                           Optimize `validate_data_columns` (#7326)
80   93ec9df13760  7304         False                           Compute proposer shuffling only once in gossip verification (#7304)
81   2aa5d5c25e22  7359         False                           Make sure to log SyncingChain ID (#7359)
82   c8224c8d5e19  7387         False                           docs: fix broken link to voluntary exit guide (#7387)
83   43c38a6fa0cc  7378         False                           Change slog to tracing in comments (#7378)
84   beb0ce68bdf6  6922         False                           Make range sync peer loadbalancing PeerDAS-friendly (#6922)
85   3d92e3663b74  6705         False                           Modularize validator store (#6705)
86   058dae064184  7405         False                           Add requires --http when using vc subcommands --http-port (#7405)
87   0f13029c7d51  7409         False                           Don't publish data columns reconstructed from RPC columns to the gossip network (#7409)
88   8dc3d23af083  7400         False                           Add a default timeout to all `BeaconNodeHttpClient` requests (#7400)
89   e90fcbe6577c  7416         False                           Add ARM binary for macOS in release (#7416)
90   4b9c16fc7175  7199         False                           Add Electra forks to basic sim tests (#7199)
91   a497ec601cae  6975         False                           Retry custody requests after peer metadata updates (#6975)
92   e0c1f27e1303  7394         False                           simulator: Persist beacon logs (#7394)
93   92391cdac665  7284         False                           update gossipsub to the latest upstream revision (#7284)
94   593390162f47  7399         False                           `peerdas-devnet-7`: update `DataColumnSidecarsByRoot` request to use `DataColumnsByRootIdentifier` (#7399)
95   5b25a48af34b  7404         False                           Siren installation improvement (#7404)
96   e051c7ca89c8  7396         False                           Siren Pectra Feature Updates (#7396)
97   0a917989b218  7370         False                           impl test random for some types (#7370)
98   807848bc7ac4  7443         False                           Next sync committee branch bug (#7443)
99   851ee2bcedfc  7454         False                           Extract get_domain for VoluntaryExit (#7454)
100  c2c7fb87a862  7460         False                           Make DAG construction more permissive (#7460)
101  b1138c28fb94  7451         False                           Add additional mergify rules to automate triaging (#7451)
102  cc6ae9d3f09c  7463         False                           Fix mergify infinite loop. (#7463)
103  1853d836b7e4  7458         False                           Added E::slots_per_epoch() to deneb time calculation (#7458)
104  c4182e362b8f  7433         False                           simulator: Write dependency logs to separate files (#7433)
105  e0ee148d6aca  7470         False                           Prevent mergify from updating labels while CI is still running. (#7470)
106  e21198c08baa  7472         False                           One more attempt to fix mergify condition. (#7472)
107  268809a53069  7471         False                           Rust clippy 1.87 lint fixes (#7471)
108  b051a5d6cc7b  7469         False                           Delete `at-most` in `lighthouse vm create` (#7469)
109  1d27855db7be  7369         False                           impl from hash256 for `ExecutionBlockHash` (#7369)
110  23ad833747b6  7417         False                           Change default EngineState to online (#7417)
111  fcfcbf9a11b3  7481         False                           Update mdlint to disable descriptive-link-text (#7481)
112  7684d1f866ab  7372         False                           ContextDeserialize and Beacon API Improvements (#7372)
113  5393d33af823  7411         False                           Silence `Uninitialized` warn log on start-up (#7411)
114  1e6cdeb88a6a  6799         False                           feat: Add docker reproducible builds (#6799)
115  50dbfdf61243  7455         False                           Some updates to Lighthouse book (#7455)
116  af87135e3020  7484         False                           Move MD059 rule to configuration file (#7484)
117  805c2dc831e6  5047         False                           Correct reward denominator in op pool (#5047)
118  7e2df6b602a1  7474         False                           Empty list `[]` to return all validators balances (#7474)
119  f06d1d034615  7495         False                           Fix blob download from checkpointz servers (#7495)
120  0688932de28d  7497         False                           Pass blobs into `ValidatorStore::sign_block` (#7497)
121  e29b607257d8  7427         False                           Move notifier and latency service to `validator_services` (#7427)
122  7759cb8f91c0  7494         False                           Update mergify rule to not evaluate PRs that are not ready for review - to reduce noise and avoid updating stale PRs. (#7494)
123  2e96e9769b99  7507         False                           Use slice.is_sorted now that it's stable (#7507)
124  a8035d7395ea  7506         False                           Enable stdout logging in rpc_tests (#7506)
125  817f14c3491a  7500         False                           Send execution_requests in fulu (#7500)
126  537fc5bde860  7459         False                           Revive network-test logs files in CI (#7459)
127  cf0f95985540  7180         False                           Improve log readability during rpc_tests (#7180)
128  ce8d0814ad71  7246         False                           Ensure logfile permissions are maintained after rotation (#7246)
129  6af8c187e0b7  7052         False                           Publish EL Info in Metrics (#7052)
130  a2797d4bbde9  7512         False                           Fix formatting errors from cargo-sort (#7512)
131  f01dc556d157  7505         False                           Update `engine_getBlobsV2` response type and add `getBlobsV2` tests (#7505)
132  e6ef644db4e8  7493         False                           Verify `getBlobsV2` response and avoid reprocessing imported data columns (#7493)
133  7c89b970afe2  7382         False                           Handle attestation validation errors (#7382)
134  8dde5bdb4413  -            -                               [NO PR MATCH]: Update mergify rules so that I can add `waiting-on-author` on a PR that's passing CI. Remove noisy comments.
135  8989ef8fb11e  7025         False                           Enable arithmetic lint in rate-limiter (#7025)
136  b7fc03437bba  -            -                               [NO PR MATCH]: Fix condition
137  9e9c51be6fef  -            -                               [NO PR MATCH]: Remove redundant `and`
138  999b04517e35  -            -                               [NO PR MATCH]: Merge pull request #7525 from jimmygchen/mergify-again
139  0ddf9a99d64a  7332         False                           Remove support for database migrations prior to schema version v22 (#7332)
140  5cda6a6f9e4b  7522         False                           Mitigate flakiness in test_delayed_rpc_response (#7522)
141  4d21846aba6b  7533         False                           Prevent `AvailabilityCheckError` when there's no new custody columns to import (#7533)
142  39744df93f0b  7393         False                           simulator: Fix `Failed to initialize dependency logging` (#7393)
143  38a5f338fad7  7529         False                           Add `console-subscriber` feature for debugging (#7529)
144  886ceb7e25e0  6882         False                           Run Assertoor tests in CI (#6882)
145  94a1446ac955  7541         False                           Fix unexpected blob error and duplicate import in fetch blobs (#7541)
146  ae30480926b6  7521         False                           Implement EIP-7892 BPO hardforks (#7521)
147  f67068e1ec53  7518         False                           Update `staking-deposit-cli` to `ethstaker-deposit-cli` (#7518)
148  cd83d8d95ddd  7544         False                           Add a name to the Tokio task (#7544)
149  357a8ccbb996  7549         False                           Checkpoint sync without the blobs from Fulu (#7549)
150  2d9fc34d4326  7540         False                           Fulu EF tests v1.6.0-alpha.0 (#7540)
151  dcee76c0dc88  7548         False                           Update key generation in validator manager (#7548)
152  9a4972053eb5  7530         False                           Add e2e sync tests to CI (#7530)
153  d457ceeaafae  7118         False                           Don't create child lookup if parent is faulty (#7118)
154  2f807e21bede  7538         False                           Add support for nightly tests (#7538)
155  e098f667380c  7570         False                           Update kurtosis config and EL images (#7570)
156  b2e8b67e3446  7566         False                           Reduce number of basic sim test nodes from 7 to 4 (#7566)
157  170cd0f5875d  7579         False                           Store the libp2p/discv5 logs when stopping local-testnet (#7579)
158  b08d49c4cb34  7559         False                           Changes for `fusaka-devnet-1` (#7559)
159  8c6abc0b69b7  7574         False                           Optimise parallelism in compute cells operations by zipping first (#7574)
160  7416d06dce8e  7561         False                           Add genesis sync test to CI (#7561)
161  076a1c3faead  7587         False                           Data column sidecar event (#7587)
162  5f208bb85829  7578         True                            Implement basic validator custody framework (no backfill) (#7578)
163  9803d69d8045  7590         False                           Implement status v2 version (#7590)
164  5472cb85008b  7582         False                           Batch verify KZG proofs for getBlobsV2 (#7582)
165  a65f78222d69  7594         False                           Drop stale registrations without reducing CGC (#7594)
166  ccd99c138c27  7588         False                           Wait before column reconstruction (#7588)
167  dc5f5af3eb53  7595         False                           Fix flaky test_rpc_block_reprocessing (#7595)
168  4fc0665ccdd6  7592         False                           Add more context to Late Block Re-orgs (#7592)
169  6135f417a2f4  7591         False                           Add data columns sidecars debug beacon API (#7591)
170  3d2d65bf8d24  7593         False                           Advertise `--advertise-false-custody-group-count` for testing PeerDAS (#7593)
171  6786b9d12a6d  7444         True                            Single attestation "Full" implementation (#7444)
172  dd985341581f  6750         True                            Hierarchical state diffs in hot DB (#6750)
173  f67084a571d1  7437         False                           Remove reprocess channel (#7437)
174  d50924677a34  7620         False                           Remove instrumenting log level (#7620)
175  11bcccb353c0  7133         True                            Remove all prod eth1 related code (#7133)
176  e34a9a0c65d5  6551         False                           Allow the `--beacon-nodes` list to be updated at runtime (#6551)
177  3fefda68e5c1  7611         False                           Send byrange responses in the correct requested range (#7611)
178  cef04ee2ee48  7462         False                           Implement `validator_identities` Beacon API endpoint (#7462)
179  fd643c310c4e  7632         False                           Un-ignore EF test for v1.6.0-alpha.1 (#7632)
180  56b2d4b5253b  7636         False                           Remove instrumenting log level  (#7636)
181  8e3c5d152413  7644         False                           Rust 1.89 compiler lint fix (#7644)
182  a0a6b9300f11  7551         False                           Do not compute sync selection proofs for the sync duty at the current slot (#7551)
183  9b1f3ed9d1a4  7652         False                           Add gossip check (#7652)
184  83cad25d9880  7657         False                           Fix Rust 1.88 clippy errors & execution engine tests (#7657)
185  522e00f48df7  7656         False                           Fix incorrect `waker` update condition (#7656)
186  6ea5f14b3988  7597         False                           feat: better error message for light_client/bootstrap endpoint (#7597)
187  2d759f78be6c  6576         False                           Fix beacon_chain metrics descriptions (#6576)
188  6be646ca1153  7666         True                            Bump DB schema to v25 (#7666)
189  e45ba846aef5  7673         False                           Increase http client default timeout to 2s in `http-api` tests. (#7673)
190  25ea8a83b77b  7667         False                           Add Michael as codeowner for store crate (#7667)
191  c1f94d9b7bf8  7669         False                           Test database schema stability (#7669)
192  257d2707182c  6612         False                           Add voluntary exit via validator manager (#6612)
193  e305cb1b921f  7661         True                            Custody persist fix (#7661)
194  41742ce2bde9  7683         False                           Update `SAMPLES_PER_SLOT` to be number of custody groups instead of data columns (#7683)
195  69c9c7038af7  7681         False                           Use prepare_beacon_proposer endpoint for validator custody registration (#7681)
196  fcc602a7872a  7646         False                           Update fulu network configs and add `MIN_EPOCHS_FOR_DATA_COLUMN_SIDECARS_REQUESTS` (#7646)
197  a459a9af98c9  7689         False                           Fix and test checkpoint sync from genesis (#7689)
198  b35854b71f04  7692         False                           Record v2 beacon blocks http api metrics separately (#7692)
199  c7bb3b00e409  7693         False                           Fix lookups of the block at `oldest_block_slot` (#7693)
200  0f895f3066a3  7695         False                           Bump default gas limit (#7695)
201  56485cc9865a  7707         False                           Remove unneeded spans that caused debug logs to appear when level is set to `info` (#7707)
202  bd8a2a8ffbaa  7023         False                           Gossip recently computed light client data (#7023)
203  7b2f138ca7e7  -            -                               [NO PR MATCH]: Merge remote-tracking branch 'origin/stable' into release-v7.1.0
204  8e55684b066f  7723         False                           Reintroduce `--logfile` with deprecation warning (#7723)
205  8b5ccacac9c0  7663         False                           Error from RPC `send_response` when request doesn't exist on the active inbound requests (#7663)
206  cfb1f7331064  7609         False                           Release v7.1.0 (#7609)
```


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

michaelsproul committed:
Make the block cache optional (#8066)

Address contention on the store's `block_cache` by allowing it to be disabled when `--block-cache-size 0` is provided, and also making this the default.


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

jimmygchen committed:
Use scoped rayon pool for backfill chain segment processing (#7924)

Part of #7866

- Continuation of #7921

In the above PR, we enabled rayon for batch KZG verification in chain segment processing. However, using the global rayon thread pool for backfill is likely to create resource contention with higher-priority beacon processor work.


  This PR introduces a dedicated low-priority rayon thread pool `LOW_PRIORITY_RAYON_POOL` and uses it for processing backfill chain segments.

This prevents backfill KZG verification from using the global rayon thread pool and competing with high-priority beacon processor tasks for CPU resources.

However, this PR by itself doesn't prevent CPU oversubscription because other tasks could still fill up the global rayon thread pool, and having an extra thread pool could make things worse. To address this we need the beacon
processor to coordinate total CPU allocation across all tasks, which is covered in:
- #7789


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

Co-Authored-By: Eitan Seri-Levi <eserilev@ucsc.edu>

jimmygchen committed:
Consolidate `reqresp_pre_import_cache`  into `data_availability_checker` (#8045)

This PR consolidates the `reqresp_pre_import_cache` into the `data_availability_checker` for the following reasons:
- the `reqresp_pre_import_cache` suffers from the same TOCTOU bug we had with `data_availability_checker` earlier, and leads to unbounded memory leak, which we have observed over the last 6 months on some nodes.
- the `reqresp_pre_import_cache` is no longer necessary, because we now hold blocks in the `data_availability_checker` for longer since (#7961), and recent blocks can be served from the DA checker.

This PR also maintains the following functionalities
- Serving pre-executed blocks over RPC, and they're now served from the `data_availability_checker` instead.
- Using the cache for de-duplicating lookup requests.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Jimmy Chen <jimmy@sigmaprime.io>

jimmygchen committed:
Rename `--subscribe-all-data-column-subnets` to `--supernode` and make it visible in help (#8083)

Rename `--subscribe-all-data-column-subnets` to `--supernode` as it's now been officially accepted in the spec. Also make it visible in help in preparation for the fusaka release.

https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/p2p-interface.md#supernodes


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen committed:
Always upload sim test logs (#8082)

This CI job failed

https://github.com/sigp/lighthouse/actions/runs/17815533375/job/50647915897

But we lost the logs because they aren't uploaded when the job fails. This PR changes the step to always upload job, even in the case of failure.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

michaelsproul committed:
Merge remote-tracking branch 'origin/stable' into unstable

michaelsproul committed:
Refine HTTP status logs (#8098)

Ensure that we don't log a warning for HTTP 202s, which are expected on the blinded block endpoints after Fulu.


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

eserilev committed:
Reduce `TARGET_BACKFILL_SLOTS` in checkpoint sync test (#8102)

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

aviggiano committed:
Use HTTPS for xdelta3 in Cargo.toml (#8094)

No issue


  Use HTTPS for dependency


Co-Authored-By: Antonio Viggiano <agfviggiano@gmail.com>

eserilev committed:
Run reconstruction inside a scoped rayon pool (#8075)

Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

Co-Authored-By: Eitan Seri-Levi <eserilev@ucsc.edu>

jimmygchen committed:
Only send data coumn subnet discovery requests after peerdas is scheduled (#8109)

#8105 (to be confirmed)

I noticed a large number of failed discovery requests after deploying latest `unstable` to some of our testnet and mainnet nodes. This is because of a recent PeerDAS change to attempt to maintain sufficient peers across data column subnets - this shouldn't be enabled on network without peerdas scheduled, otherwise it will keep retrying discovery on these subnets and never succeed.

Also removed some unused files.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Jimmy Chen <jimmy@sigmaprime.io>

dapplion committed:
Only mark block lookups as pending if block is importing from gossip (#8112)

- PR https://github.com/sigp/lighthouse/pull/8045 introduced a regression of how lookup sync interacts with the da_checker.

Now in unstable block import from the HTTP API also insert the block in the da_checker while the block is being execution verified. If lookup sync finds the block in the da_checker in `NotValidated` state it expects a `GossipBlockProcessResult` message sometime later. That message is only sent after block import in gossip.

I confirmed in our node's logs for 4/4 cases of stuck lookups are caused by this sequence of events:
- Receive block through API, insert into da_checker in fn process_block in put_pre_execution_block
- Create lookup and leave in AwaitingDownload(block in processing cache) state
- Block from HTTP API finishes importing
- Lookup is left stuck

Closes https://github.com/sigp/lighthouse/issues/8104


  - https://github.com/sigp/lighthouse/pull/8110 was my initial solution attempt but we can't send the `GossipBlockProcessResult` event from the `http_api` crate without adding new channels, which seems messy.

For a given node it's rare that a lookup is created at the same time that a block is being published. This PR solves https://github.com/sigp/lighthouse/issues/8104 by allowing lookup sync to import the block twice in that case.


Co-Authored-By: dapplion <35266934+dapplion@users.noreply.github.com>

eserilev committed:
Fulu testnet configs (#8117)

Holesky - #8096
Hoodi - #8097
Sepolia - #8099


  Testnet configs for Holesky, Hoodi and Sepolia

Holesky - https://github.com/eth-clients/holesky/pull/132
Hoodi - https://github.com/eth-clients/hoodi/pull/21
Sepolia - https://github.com/eth-clients/sepolia/pull/111


Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

michaelsproul committed:
Fix bugs in proposer calculation post-Fulu (#8101)

As identified by a researcher during the Fusaka security competition, we were computing the proposer index incorrectly in some places by computing without lookahead.


  - [x] Add "low level" checks to computation functions in `consensus/types` to ensure they error cleanly
- [x] Re-work the determination of proposer shuffling decision roots, which are now fork aware.
- [x] Re-work and simplify the beacon proposer cache to be fork-aware.
- [x] Optimise `with_proposer_cache` to use `OnceCell`.
- [x] All tests passing.
- [x] Resolve all remaining `FIXME(sproul)`s.
- [x] Unit tests for `ProtoBlock::proposer_shuffling_root_for_child_block`.
- [x] End-to-end regression test.
- [x] Test on pre-Fulu network.
- [x] Test on post-Fulu network.


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

pawanjay176 committed:
Fix bug in fork calculation at fork boundaries (#8121)

N/A


  In #8101 , when we modified the logic to get the proposer index post fulu, we seem to have missed advancing the state at the fork boundaries to get the right `Fork` for signature verification.
This led to lighthouse failing all gossip verification right after transitioning to fulu that was observed on the holesky shadow fork
```
Sep 26 14:24:00.088 DEBUG Rejected gossip block                         error: "InvalidSignature(ProposerSignature)", graffiti: "grandine-geth-super-1", slot: 640
Sep 26 14:24:00.099 WARN  Could not verify block for gossip. Rejecting the block  error: InvalidSignature(ProposerSignature)
```

I'm not completely sure this is the correct fix, but this fixes the issue with `InvalidProposerSignature` on the holesky shadow fork.

Thanks to @eserilev for helping debug this


Co-Authored-By: Pawan Dhananjay <pawandhananjay@gmail.com>

michaelsproul committed:
Fix proposer shuffling decision slot at boundary (#8128)

Follow-up to the bug fixed in:

- https://github.com/sigp/lighthouse/pull/8121

This fixes the root cause of that bug, which was introduced by me in:

- https://github.com/sigp/lighthouse/pull/8101

Lion identified the issue here:

- https://github.com/sigp/lighthouse/pull/8101#discussion_r2382710356


  In the methods that compute the proposer shuffling decision root, ensure we don't use lookahead for the Fulu fork epoch itself. This is accomplished by checking if Fulu is enabled at `epoch - 1`, i.e. if `epoch > fulu_fork_epoch`.

I haven't updated the methods that _compute_ shufflings to use these new corrected bounds (e.g. `BeaconState::compute_proposer_indices`), although we could make this change in future. The `get_beacon_proposer_indices` method already gracefully handles the Fulu boundary case by using the `proposer_lookahead` field (if initialised).


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

jimmygchen committed:
Release v8.0.0 rc.0 (#8127)

Testnet release for the upcoming Fusaka fork.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Jimmy Chen <jimmy@sigmaprime.io>

michaelsproul committed:
Update book for DB schema v28 (#8132)

Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

macladson committed:
Bump superstruct to `0.10.0` (#8133)

Bump `superstruct` to the latest release `0.10.0`.
This version uses a later version of `darling` which is helpful for https://github.com/sigp/lighthouse/pull/8125


Co-Authored-By: Mac L <mjladson@pm.me>

michaelsproul committed:
Improve spec compliance for `/eth/v1/config/spec` API (#8144)

- [x] Remove the unnecessary `_MILLIS` suffix from `MAXIMUM_GOSSIP_CLOCK_DISPARITY`
- [x] Add missing Deneb preset `KZG_COMMITMENT_INCLUSION_PROOF_DEPTH`, not to be confused with `KZG_COMMITMENTS_INCLUSION_PROOF_DEPTH` (plural) from Fulu...


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

jimmygchen committed:
Remove unnecessary warning logs and update logging levels (#8145)

@michaelsproul noticed this warning on a devnet-3 node

```
Oct 01 16:37:29.896 WARN  Error when importing rpc custody columns      error: ParentUnknown { parent_root: 0xe4cc85a2137b76eb083d7076255094a90f10caaec0afc8fd36807db742f6ff13 }, block_hash: 0x43ce63b2344990f5f4d8911b8f14e3d3b6b006edc35bbc833360e667df0edef7
```

We're also seeing similar `WARN` logs for blobs on our live nodes.

It's normal to get parent unknown in lookups and it's handled here
https://github.com/sigp/lighthouse/blob/a134d43446f776fe2a84f420854afbff76ca93d8/beacon_node/network/src/sync/block_lookups/mod.rs#L611-L619

These shouldn't be a `WARN`, and we also log the same error in block lookups at `DEBUG` level here:
https://github.com/sigp/lighthouse/blob/a134d43446f776fe2a84f420854afbff76ca93d8/beacon_node/network/src/sync/block_lookups/mod.rs#L643-L648

So i've removed these extra WARN logs.

I've also lower the level of an `ERROR` log when unable to serve data column root requests - it's unexpected, but is unlikely to impact the nodes performance, so I think we can downgrade this.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

eserilev committed:
Fulu ASCII art (#8151)

Co-Authored-By: Eitan Seri- Levi <eserilev@gmail.com>

pawanjay176 committed:
Fallback to getPayload v1 if v2 fails (#8163)

N/A


  Post fulu, we should be calling the v2 api on the relays that doesn't return the blobs/data columns.

However, we decided to start hitting the v2 api as soon as fulu is scheduled to avoid unexpected surprises at the fork.
In the ACDT call, it seems like most clients are calling v2 only after the fulu fork.
This PR aims to be the best of both worlds where we fallback to hitting v1 api if v2 fails. This way, we know beforehand if relays don't support it and can potentially alert them.


Co-Authored-By: Pawan Dhananjay <pawandhananjay@gmail.com>

michaelsproul committed:
Quote `BeaconState::proposer_lookahead` in JSON repr (#8167)

Use quoted integers for `state.proposer_lookahead` when serializing JSON. This is standard for all integer fields, but was missed for the newly added proposer lookahead. I noticed this issue while inspecting the head state on a local devnet.

I'm glad we found this before someone reported it :P


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

jimmygchen committed:
Remove deprecated CLI flags and references for v8.0.0 (#8142)

Closes #8131

- [x] Remove deprecated flags from beacon_node/src/cli.rs:
- [x] eth1-purge-cache
- [x] eth1-blocks-per-log-query
- [x] eth1-cache-follow-distance
- [x] disable-deposit-contract-sync
- [x] light-client-server
- [x] Remove deprecated flags from lighthouse/src/main.rs:
- [x] logfile
- [x] terminal-total-difficulty-override
- [x] terminal-block-hash-override
- [x] terminal-block-hash-epoch-override
- [x] safe-slots-to-import-optimistically
- [x] Remove references to deprecated flags in config.rs files
- [x] Remove warning messages for deprecated flags in main.rs
- [x] Update/remove related tests in beacon_node.rs


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

Co-Authored-By: Jimmy Chen <jimmy@sigmaprime.io>

michaelsproul committed:
Block proposal optimisations (#8156)

Closes:

- https://github.com/sigp/lighthouse/issues/4412

This should reduce Lighthouse's block proposal times on Holesky and prevent us getting reorged.


  - [x] Allow the head state to be advanced further than 1 slot. This lets us avoid epoch processing on hot paths including block production, by having new epoch boundaries pre-computed and available in the state cache.
- [x] Use the finalized state to prune the op pool. We were previously using the head state and trying to infer slashing/exit relevance based on `exit_epoch`. However some exit epochs are far in the future, despite occurring recently.


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

pawanjay176 committed:
Bump kzg library versions (#8174)

N/A


  Update c-kzg and rust-eth-kzg to their latest versions. Also removes the patch version hardcoding in Cargo.toml.


Co-Authored-By: Pawan Dhananjay <pawandhananjay@gmail.com>

chong-he committed:
Implement `/eth/v1/beacon/blobs` endpoint (#8103)

* #8085


  


Co-Authored-By: Tan Chee Keong <tanck@sigmaprime.io>

Co-Authored-By: chonghe <44791194+chong-he@users.noreply.github.com>

jimmygchen committed:
Reject data columns that does not descend from finalize root instead of ignoring it (#8179)

This issue was identified during the fusaka audit competition.

The [`verify_parent_block_and_finalized_descendant`](https://github.com/sigp/lighthouse/blob/62d9302e0f9dd9f94d0325411a3029b36ad90685/beacon_node/beacon_chain/src/data_column_verification.rs#L606-L627) in data column gossip verification currently load the parent first before checking if the column descends from the finalized root.

However, the `fork_choice.get_block(&block_parent_root)` function also make the same check internally:

https://github.com/sigp/lighthouse/blob/8a4f6cf0d5b6b261b2c3439ce7c05383a53d30c5/consensus/fork_choice/src/fork_choice.rs#L1242-L1249

Therefore, if the column does not descend from the finalized root, we return an `UnknownParent` error, before hitting the `is_finalized_checkpoint_or_descendant` check just below.

Which means we `IGNORE` the gossip message instead `REJECT`, and the gossip peer is not _immediately_ penalised. This deviates from the spec.

However, worth noting that lighthouse will currently attempt to request the parent from this peer, and if the peer is not able to serve the parent, it gets penalised with a `LowToleranceError`, and will get banned after ~5 occurences.

https://github.com/sigp/lighthouse/blob/ffa7b2b2b9e3b4e70678e2c749b8bc45234febd7/beacon_node/network/src/sync/network_context.rs#L1530-L1532

This PR will penalise the bad peer immediately instead of performing block lookups before penalising it.


  


Co-Authored-By: Jimmy Chen <jchen.tc@gmail.com>

michaelsproul committed:
Update `ForkName::latest_stable` to Fulu for tests (#8181)

Update `ForkName::latest_stable` to Fulu, reflecting our plan to stabilise Fulu in the immediate future!

This will lead to some more tests running with Fulu rather than Electra.


Co-Authored-By: Michael Sproul <michael@sigmaprime.io>

michaelsproul commented:
@sashaodessa Can you please update this PR with the latest changes from `unstable`?

sashaodessa commented:
> @sashaodessa Can you please update this PR with the latest changes from `unstable`?

@michaelsproul Hi, I flooded my MacBook and took it to a repair shop. Unfortunately, I only have my phone now. I`ll try from another device

sashaodessa committed:
Merge branch 'stable' into unstable

michaelsproul commented:
This diff is way too large now
